geopandas/vectorized.pyx

```diff
@@ -0,0 +1,1029 @@
+    elif op == 'union':
+        func = GEOSUnion_r
+    elif op == 'intersection':
+        func = GEOSIntersection_r
+    else:
+        raise NotImplementedError("Op %s not known" % op)
+
+    with nogil:
+        for idx in xrange(n):
+            left_geom = <GEOSGeometry *> left[idx]
+            if left_geom and right_geom:
+                out[idx] = <np.uintp_t> func(handle, left_geom, right_geom)
+            else:
+                out[idx] = 0
+
+    return VectorizedGeometry(out)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef buffer(np.ndarray[np.uintp_t, ndim=1, cast=True] geoms, double distance,
+            int resolution, int cap_style, int join_style, double mitre_limit):
+    cdef Py_ssize_t idx
+    cdef GEOSContextHandle_t handle
+    cdef GEOSGeometry *geom
+    cdef uintptr_t geos_geom
+    cdef GEOSGeometry *other_geom
+    cdef unsigned int n = geoms.size
+
+    cdef np.ndarray[np.uintp_t, ndim=1, cast=True] out = np.empty(n, dtype=np.uintp)
+    handle = get_geos_context_handle()
+
+    with nogil:
+        for idx in xrange(n):
+            geos_geom = geoms[idx]
+            geom = <GEOSGeometry *> geos_geom
+            if geom:
+                out[idx] = <np.uintp_t> GEOSBufferWithStyle_r(handle, geom,
+                        distance, resolution, cap_style, join_style, mitre_limit)
+            else:
+                out[idx] = 0
+
+    return VectorizedGeometry(out)
+
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef get_coordinate_point(np.ndarray[np.uintp_t, ndim=1, cast=True] geoms,
+                          int coordinate):
+    cdef Py_ssize_t idx
+    cdef GEOSContextHandle_t handle
+    cdef GEOSGeometry *geom
+    cdef GEOSCoordSequence *sequence
+    cdef unsigned int n = geoms.size
+    cdef double value
+
+    cdef np.ndarray[double, ndim=1, cast=True] out = np.empty(n, dtype=np.float)
+
+    handle = get_geos_context_handle()
+
+    if coordinate == 0:
+        func = GEOSCoordSeq_getX_r
+    elif coordinate == 1:
+        func = GEOSCoordSeq_getY_r
+    elif coordinate == 2:
+        func = GEOSCoordSeq_getZ_r
+    else:
+        raise NotImplementedError("Coordinate must be between 0-x, 1-y, 2-z")
+
+    with nogil:
+        for idx in xrange(n):
+            geom = <GEOSGeometry *> geoms[idx]
+            if geom:
+                sequence = GEOSGeom_getCoordSeq_r(handle, geom)
+                func(handle, sequence, 0, &value)
+                out[idx] = value
+            else:
+                out[idx] = 0
+
+    return out
+
+
+cpdef from_shapely(object L):
+    cdef Py_ssize_t idx
+    cdef GEOSContextHandle_t handle
+    cdef GEOSGeometry *geom
+    cdef uintptr_t geos_geom
+    cdef unsigned int n
+
+    n = len(L)
+
+    cdef np.ndarray[np.uintp_t, ndim=1] out = np.empty(n, dtype=np.uintp)
+
+    handle = get_geos_context_handle()
+
+    for idx in xrange(n):
+        g = L[idx]
+        if g is not None and not (isinstance(g, float) and np.isnan(g)):
```

Comment:
I added the `(isinstance(g, float) and np.isnan(g))` here, but not sure this is the best way to check for NaN